M2/Macaulay2/m2/matrix2.m2

```diff
@@ -247,6 +247,26 @@ trimHelper = ((opts, M) -> (
 addHook((trim, Module), Strategy => Inhomogeneous, (opts, M) -> trimHelper(opts ++ {Strategy => Inhomogeneous}, M))
 addHook((trim, Module), Strategy => Complement,    (opts, M) -> trimHelper(opts ++ {Strategy => Complement},    M))
 
+trimPID := M -> if M.?relations then (if M.?generators then trimPID image generators M else ambient M) / trimPID image relations M else if not M.?generators then M else (
+    f := presentation M;
+    (g,ch) := smithNormalForm(f, ChangeMatrix => {true, false});
+    isunit := r -> r != 0 and degree r === {0};
+    rows := select(min(rank source f,rank target f),i->isunit g_(i,i));
+    rows = rows | toList(rank target f..<rank target g); -- temporary fix for #3017
+    ch = submatrix'(ch,rows,);
+    p:=id_(target ch)//ch;
+    q:=generators M*p;
+    d:=diagonalMatrix apply(rank source q,i->lcm apply(entries (q_i),a->if a==0 then 1 else 1/leadCoefficient a));
+    image (q*d)
+    )
+addHook((trim, Module), Strategy => "PID",
+    (opts, M) -> (
+	R := ring M;
+	if instance(R,PolynomialRing) and numgens R === 1 and isField coefficientRing R and not isHomogeneous M then trimPID M
+	)
```

Comment:
Rather than repeating the prune algorithm in trimPID, how about this instead?
```suggestion
	then if isFreeModule M then M else subquotient(
	    if M.?generators   then (prune image generators M).cache.pruningMap,
	    if M.?relations    then (prune image relations  M).cache.pruningMap))
```

Reply:
I don't think what you wrote works. anyway I did have an earlier version going through `prune`, but it was fairly complicated. I think the current code is cleaner.

Reply:
Sorry, what doesn't work?

Reply:
I tried your code with `coker matrix {{x}}` and it didn't produce the expected result.

Reply:
What's the expected result, not this?!
```m2
i1 : R = QQ[x]; trim coker matrix {{x}}

o2 = cokernel | x |
```

Reply:
You must have used `M.relations` for both inputs of `subquotient`.

Reply:
no, I copy/pasted your code:
```
i1 : test = M -> if isFreeModule M then M else subquotient(
	    if M.?generators   then (prune image generators M).cache.pruningMap,
	    if M.?relations    then (prune image relations  M).cache.pruningMap)
          
o1 = test

o1 : FunctionClosure

i2 : R=QQ[x];

i3 : test cokernel matrix {{x}}

o3 = subquotient (| x |, | x |)

                               1
o3 : R-module, subquotient of R
```

Reply:
also it's about twice as slow due lots of junk in the `prune` code.

Reply:
These are valid improvements to the PID strategy for prune that I think should be implemented, but the main computation at the center of trim and prune for an ideal or module over PID is exactly the same, and duplicating an algorithm in two places is a recipe for bugs and inconsistencies and harder maintenance in the future.

If you want me to suggest a different change that doesn't have the slow down you describe (pull the prune strategy into a helper function and use it both places) or the denominators (divide by the leading coefficients) I can, but this is your PR and since my impression is that you also don't like hard to parse code I wanted to suggest a way to avoid this mess:
```m2
trimPID := M -> if M.?relations then (if M.?generators then trimPID image generators M else ambient M) / trimPID image relations M else if not M.?generators then M else (
```

If you don't care about code duplication and inconsistency between trim and prune then never mind.

Reply:
I'll try to rewrite to avoid redundancies and ifs